lang/src/lib.rs

```diff
@@ -357,12 +355,53 @@ pub mod __private {
 macro_rules! require {
     ($invariant:expr, $error:tt $(,)?) => {
         if !($invariant) {
-            return Err(crate::ErrorCode::$error.into());
+            return Err(anchor_lang::anchor_attribute_error::error!(
+                crate::ErrorCode::$error
+            ));
         }
     };
     ($invariant:expr, $error:expr $(,)?) => {
         if !($invariant) {
-            return Err($error.into());
+            return Err(anchor_lang::anchor_attribute_error::error!($error));
+        }
+    };
+}
+
+/// Returns with the given error.
+/// Use this with a custom error type.
+///
+/// # Example
+/// ```ignore
+/// // Instruction function
+/// pub fn example(ctx: Context<Example>) -> Result<()> {
+///     bail!(MyError::SomeError);
+/// }
+///
+/// // An enum for custom error codes
+/// #[error_code]
+/// pub enum MyError {
+///     SomeError
+/// }
+/// ```
+#[macro_export]
+macro_rules! bail {
```

Comment:
Bikeshed: Kind of think we should rename this to `err!`, which is more descriptive. I'm not sure anyhow should be considered the authority rust error handling.

Reply:
Also think we should suggest using it. `err!(ErrorCode)` is much nicer than `Err(error!(ErrorCode))`

Reply:
`err!` implies `Err(error!(ErrorCode))` though which is why I used `bail!` to indicate that it expands to `return Err(error!(ErrorCode));`

Reply:
I see the difference now and why bail is actually better, given the implicit return.. Personally, I'm not a big fan of the implicit return inside the macro. Would rather see `return err!(ErrorCode)`.

Reply:
yea I like that more as well